### PR TITLE
Remove warning

### PR DESCRIPTION
**Pull request by filipecabaco**
Bump elixir version and remove Access protocol to remove

`lib/httpotion.ex:292: warning: implementation of the Access protocol is deprecated. For customization of the dict[key] syntax, please implement the Dict behaviour instead`

Reference: https://github.com/elixir-lang/elixir/issues/2973

### CODE REVIEW

**Comment by myfreeweb** [comment]
You can't just remove it!! That would break everything that reads header values!

Elixir completely fucked up the `[]` overloading thing :( Access deprecation warning says "implement Dict", but [Dict itself is deprecated](http://elixir-lang.org/docs/stable/elixir/Dict.html)?!

**Comment by filipecabaco** [comment]
Ok, assumed that if the tests were fine it would be ok, clearly missed it :\ 

Sorry for the issue and PR

**Comment by myfreeweb** [comment]
Wait. What the hell. Tests did pass (on Travis).

There are lines like `assert response.headers[:"Content-Length"] == "0"` in the tests.

Did... the `[]` access work without the `Access` protocol?

**Comment by myfreeweb** [comment]
No. No way. On my machine, just removing the `Access` impl results in this:
```
Compiled lib/httpotion.ex
** (EXIT from #PID<0.47.0>) an exception was raised:
    ** (MatchError) no match of right hand side value: {:error, :not_a_protocol}
        (mix) lib/mix/tasks/compile.protocols.ex:108: Mix.Tasks.Compile.Protocols.consolidate/3
        (elixir) lib/task/supervised.ex:89: Task.Supervised.do_apply/2
        (elixir) lib/task/supervised.ex:40: Task.Supervised.reply/5
        (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3

01:24:05.511 [error] Task #PID<0.149.0> started from #PID<0.47.0> terminating
** (MatchError) no match of right hand side value: {:error, :not_a_protocol}
    (mix) lib/mix/tasks/compile.protocols.ex:108: Mix.Tasks.Compile.Protocols.consolidate/3
    (elixir) lib/task/supervised.ex:89: Task.Supervised.do_apply/2
    (elixir) lib/task/supervised.ex:40: Task.Supervised.reply/5
    (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
Function: #Function<6.51313027/0 in Mix.Tasks.Compile.Protocols.consolidate/5>
    Args: []
```

HOW DID THIS PASS ON TRAVIS WHAT IS GOING ON

**Comment by filipecabaco** [comment]
Yes, I actually tried it on iex just to be extra sure that we would be able to extract  and tried  ` (HTTPotion.get "httpbin.org/get").headers[:date] ` and worked without a problem that is why I assumed this change would be fine.

Not sure :\

**Comment by myfreeweb** [comment]
Holy shit. It's the `elixir:  "~> 1.2"` thing. It does work with that.

I had NO IDEA that the `elixir` version in `mix.exs` actually *changed the behavior of the language*. What the fuck.

**Comment by myfreeweb** [comment]
Wait. Did Mix actually *use Elixir 1.0 for compiling* because of `elixir: "~> 1.0"`?? What is going on

**Comment by filipecabaco** [comment]
tbh I can't find any good documentation regarding the behaviour of the `elixir` key. 

But actually I was able to do it even without the version bump... Maybe a newer version of elixir is doing something weird? not sure...

Versions I'm using just for reference:  ```Hex:    0.11.5
 Elixir: 1.2.4
 OTP:    18.3```

**Comment by myfreeweb** [comment]
@filipecabaco This whole situation is all about Elixir's (lack of) documentation… Where is the official explanation of overloading [] without implementing Access?

**Comment by josevalim** [comment]
@myfreeweb the Access issue is a documentation issue and we will improve the warning.

The `:elixir` key does not change the behaviour of the language. It does not pick an Elixir version, it uses whatever is in your system. The key specifies which Elixir version is acceptable for the project.

In Elixir v1.0, Access used protocol dispatch. In Elixir v1.1 onwards, if you call `headers[:foo]`, It will dispatch to `Headers.fetch(headers, :foo)` (which is the Dict contract), which you have implemented and that's why it works.

If something is wrong or incomplete, please report an issue. If something is confusing, it is likely the language's fault, but we need to know that such is happening otherwise it is impossible for us to improve the parts that are lacking in the language.

TL;DR: Removing the Access implementation in your case should just work.

**Comment by myfreeweb** [comment]
@josevalim Thanks! This is the first time I've heard of "the Dict contract" as opposed to "the Dict protocol" which is deprecated. Is the contract documented anywhere?

I still don't understand [the thing happening here](https://asciinema.org/a/4i828yv37p41omw0xunt2r78r). I'm getting the `(MatchError) no match of right hand side value: {:error, :not_a_protocol}` message when the `:elixir` key is set to `~> 1.0` but not `~> 1.2`. How is that error possible?

**Comment by josevalim** [comment]
It is a documentation issue because it is not supposed to be "Dict
behavior" but "Access behavior". It is documented in the Access module.

I have cloned your repository and I will try to reproduce the
not_a_protocol error. My suspicion is that it is related to stale files in
_build. Mix was supposed to remove them but it seems it did not. If you
nuke _build, it should work. What is your local elixir version?



-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by myfreeweb** [comment]
Elixir 1.2.0, Mix 1.2.0.

Yeah looks like I can't reproduce it anymore after changing versions.
